Avoid printing confusing error messages.

jwpeterson opened the pull request:
Instead of trying to handle every possible case, e.g. forgetting to specify a variable, or setting `variable = ''`, or `variable = ' '`, we can just alert the user that the variable parameter is either missing or not a valid nonlinear variable.

Add new test which attempts to add a Kernel with no variable parameter. The error message is now:
```
*** ERROR ***
Error adding Kernel diff, "variable" parameter either missing or does not specify a valid nonlinear variable!
```
Also update the expect_err message for related check_error tests.

Refs #11192.

jwpeterson committed:
Avoid printing confusing error messages.

Instead of trying to handle every possible case, e.g. forgetting to
specify a variable, or setting variable = '', we can just alert the
user that the parameter is either missing or invalid for some reason.

Add new test which attempts to add a Kernel with no variable
parameter. The new error message is:

*** ERROR ***
Error adding Kernel diff, "variable" parameter either missing or does not specify a valid nonlinear variable!

Also update the expect_err message for related check_error tests.

Refs #11192.

jwpeterson commented:
@dschwen if the idea works for you, I will also need to update the other `FEProblemBase::addXYZ()` functions in a similar manner...

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/175072/) on 6eb9ac0 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](http://mooseframework.inl.gov/docs/PRs/11227/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s http://mooseframework.inl.gov/docs/PRs/11227/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format c23f29592a0205585ecebebc989bd2cccefc5ab4`

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/175088/) on 26eb2df wanted to post the following:

View the site [here](http://mooseframework.inl.gov/docs/PRs/11227/site)

This comment will be updated on new commits.

friedmud commented:
I'm a bit confused - is `variable` _not_ a required parameter for Kernel?

jwpeterson commented:
It is... the original issue was that people were setting Aux variables on Kernels and we weren't catching that, necessitating a new error check.

moosebuild commented:
Job [Modules debug](https://civet.inl.gov/job/175084/) on 26eb2df : invalidated by @jwpeterson

Apparently a random MPI error on buildq21.

jwpeterson committed:
Add new args to MooseVariableInterface constructor.

This is intended to allow additional error checking and catch cases in
which someone incorrectly assigns e.g. an AuxVariable to a Kernel or a
NonlinearVariable to an AuxKernel.

* Add SubProblem::getVariableWithChecks() interface and derived class implementations.
* Drop hasVariable() checks before FEProblemBase::addXYZ() calls.
* Update expected error messages in misc/check_error tests.
* Add check (and test) to catch rare case where variable name is set to empty string.
* Move error message into default case to avoid warning.
  The new enum caused the warning:
  enumeration value 'VAR_ANY' not handled in switch [-Wswitch]" warning in debug mode.
* Add MooseVariableBase::isVector().
* Add Moose::VarFieldType enumeration.
* Add logic checks in FEProblemBase::getVariableWithChecks().
* Add logic checks in DisplacedProblem::getVariableWithChecks().
* Add test verifying that we can't add a standard Kernel with a vector variable.

Refs #11192.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/176746/) on 0f953fe wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](http://mooseframework.inl.gov/docs/PRs/11227/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s http://mooseframework.inl.gov/docs/PRs/11227/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 6fa9ee29eea9380872670281093fd496e4a8c5ca`

moosebuild commented:
Job [Modules debug](https://civet.inl.gov/job/176769/) on 1e0a338 : invalidated by @jwpeterson

level_set/test:reinitialization.full timed out. Maybe this test is too slow in debug mode...

jwpeterson commented:
@friedmud the checks you recommended doing through `MooseVariableInterface` are now implemented, and the `FEProblemBase` error checking is removed.

jwpeterson committed:
Add error checking args to SubProblem::getVariable().

The additional arguments have sensible default values so the change
should be backwards compatible. I also went through the framework,
modules and apps to figure out which flags were appropriate to
pass in various getVariable calls. The following classes were updated
in modules:

* CrackFrontDefinition
* CrackFrontData
* RichardsMultiphaseProblem
* LevelSetBiMaterialBase
* LevelSetCutUserObject

Refs #11192.

A note on default values for virtual functions:

According to the standard, default values are determined by the
_static_ type used to make the call to the function. Since
getVariable() is frequently called through an FEProblem reference, its
getVariable() override must provide the defaults. A good rule of thumb
is: "Make default arguments the same or absent when overriding a
virtual function."

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/178555/) on 2bc9aed wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](http://mooseframework.inl.gov/docs/PRs/11227/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s http://mooseframework.inl.gov/docs/PRs/11227/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 6fa9ee29eea9380872670281093fd496e4a8c5ca`

jwpeterson committed:
New default args should go at end of parameter list.

This is the most backwards-compatible approach, although I did not
come across any apps that were broken by the other order.  Also changed
the default values to "ANY" for both, as this is also the most backwards
compatible behavior.

Use the new constructor for MultiAppDTKUserObjectTransfer, apparently
I forgot to do that one before...

friedmud commented on framework/src/variables/MooseVariableBase.C:
```diff
@@ -56,3 +56,9 @@ MooseVariableBase::order() const
+bool
+MooseVariableBase::isVector() const
+{
+  return (_fe_type.family == LAGRANGE_VEC || _fe_type.family == NEDELEC_ONE);
```
Worried about needing to remember to update this if stuff changes.

What do you think about making this virtual - and letting the MooseVariable derived types implement it to be true/false.

@lindsayad do you have a comment on that?

lindsayad replied:
It would be straightforward to implement it in the derived classes. @jwpeterson and I talked briefly about this. One implementation would be:
```c++
template <typename T>
bool
MooseVariableFEImpl<T>::isVector() const
{
  return std::is_same<T, RealVectorValue>::value;
}
```

jwpeterson replied:
Yeah I know this isn't ideal.  We could add template specializations of `FE<Dim,Family>::is_vector()` that returns true/false at the libmesh level, but I don't think that would help since we really just have an `FEType` at the `MooseVariableBase` level...

friedmud commented on framework/src/variables/MooseVariableInterface.C:
```diff
@@ -31,11 +33,22 @@ MooseVariableInterface<T>::MooseVariableInterface(const MooseObject * moose_obje
+    // in their input file. This could happen if e.g. something goes
+    // wrong with dollar bracket expression expansion.
+    if (vec.empty())
+      mooseError("Error retrieving variable name! Did you set variable = '' by accident?");
```
We're trying to make errors a little less opaque (lol - that's actually the function of _this_ PR!): would you mind adding as much context as you can to this error statement?

jwpeterson replied:
Hmm... maybe we could print `name` of the passed in MooseObject's `InputParameters`?

friedmud replied:
Just do what you can - any amount of context will help

friedmud commented on framework/include/utils/MooseTypes.h:
```diff
@@ -186,7 +186,16 @@ enum AuxGroup
+
+enum VarFieldType
+{
+  VAR_FIELD_STANDARD,
```
@lindsayad does this naming match up with your new naming for MooseVariable types?

friedmud reviewed (changes requested):
This is _really_ good.  This was seriously needed and will  really help users diagnose problems.

I only have a few small questions...

jwpeterson committed:
Fix issues related to MooseVariableFE -> MooseVariableFEBase name change.

jwpeterson committed:
Use Transient Executioner in check_error.scalar_kernel_with_var test.

This lets us avoid an unrelated error from the Tagging system and get
the error we are actually trying to test.

jwpeterson committed:
Make isVector() virtual, use std::is_same implementation.

jwpeterson committed:
Improve error message when user specifies empty string for variable name.

jwpeterson committed:
Add SubProblem::getVariableHelper() function.

This allows us to share code between FEProblemBase and DisplacedProblem rather than duplicating it.

jwpeterson committed:
Apply clang-format changes.

jwpeterson commented:
@friedmud I believe I have addressed all your comments.

friedmud reviewed (approved):
👍
